// @scalar/openapi-parser/dist/index.js
import { dereference } from "./utils/dereference.js";
import { escapeJsonPointer } from "./utils/escape-json-pointer.js";
import { filter } from "./utils/filter.js";
import { isJson } from "./utils/is-json.js";
import { isYaml } from "./utils/is-yaml.js";
import { load } from "./utils/load/index.js";
import { normalize } from "./utils/normalize.js";
import { sanitize } from "./utils/transform/sanitize.js";
import { toJson } from "./utils/to-json.js";
import { toYaml } from "./utils/to-yaml.js";
import { traverse } from "./utils/traverse.js";
import { unescapeJsonPointer } from "./utils/unescape-json-pointer.js";
import { upgrade } from "./utils/upgrade.js";
import { upgradeFromThreeToThreeOne } from "./utils/upgrade-from-three-to-three-one.js";
import { upgradeFromTwoToThree } from "./utils/upgrade-from-two-to-three.js";
import { validate } from "./utils/validate.js";
import { join } from "./utils/join/index.js";
import { openapi } from "./utils/openapi/openapi.js";
export {
  dereference,
  escapeJsonPointer,
  filter,
  isJson,
  isYaml,
  join,
  load,
  normalize,
  openapi,
  sanitize,
  toJson,
  toYaml,
  traverse,
  unescapeJsonPointer,
  upgrade,
  upgradeFromThreeToThreeOne,
  upgradeFromTwoToThree,
  validate
};
